packages/mui-material/src/Modal/ModalManager.ts

```diff
@@ -56,22 +56,56 @@ function isAriaHiddenForbiddenOnElement(element: Element): boolean {
   return isForbiddenTagName || isInputHidden;
 }
 
-function ariaHiddenSiblings(
+function ariaHiddenElements(
   container: Element,
-  mountElement: Element,
+  mountElement: Element | null,
   currentElement: Element,
   elementsToExclude: readonly Element[],
   show: boolean,
 ): void {
-  const blacklist = [mountElement, currentElement, ...elementsToExclude];
+  let current: Element | null = container;
+  let previousElement: Element =
+    container === mountElement ? currentElement : (mountElement ?? currentElement);
+  const html = ownerDocument(container).body.parentElement;
+  const blacklist = [mountElement, ...elementsToExclude];
+
+  // In theory this should not happen anymore.
+  // in some cases the container and previous element still
+  // could end up being the same, in this case we just go up 1
+  if (current === previousElement) {
+    current = current.parentElement;
+  }
 
-  [].forEach.call(container.children, (element: Element) => {
-    const isNotExcludedElement = !blacklist.includes(element);
-    const isNotForbiddenElement = !isAriaHiddenForbiddenOnElement(element);
-    if (isNotExcludedElement && isNotForbiddenElement) {
-      ariaHidden(element, show);
+  while (!!current && html !== current) {
+    for (let i = 0; i < current.children.length; i += 1) {
+      const element = current.children[i];
+      const isNotExcludedElement = blacklist.indexOf(element) === -1;
+      const isNotForbiddenElement = !isAriaHiddenForbiddenOnElement(element);
+      const isPreviousElement = element === previousElement;
+
+      // We came from here
+      if (isPreviousElement) {
+        if (!isNotExcludedElement) {
+          // If any of the ancestors have aria-hidden applied (e.g. by another Modal)
+          // there is a chance that we end up with nothing accessible in the element tree.
+          // So we remove the aria-hidden tag from ancestors so at least the current modal is accessible,
+          // even tho it's probably undesirable when aria-hidden is not coming from another modal.
+          if (show) {
+            ariaHidden(element, !show);
+          }
+          // we restore it if it was originally hidden
+          else {
+            ariaHidden(element, show);
+          }
+        }
```

Comment:
Can you explain this part of the logic with an example? It's a little hard to follow.

Reply:
Before this function it only hid siblings. Now it traverses up the html tree from the modal mount location, hiding every sibling and removing aria-hidden from every ancestor of the modal (if needed).

This because models can be mounted anywhere, and we always want to only allow the latest open modal to be accessible.